rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java

```diff
@@ -0,0 +1,262 @@
+
+        RedeemData redeemData = RedeemData.of(p2shErpFederation.getBtcPublicKeys(), flyoverP2shErpRedeemScript);
+        Script flyoverP2sh = ScriptBuilder.createP2SHOutputScript(flyoverP2shErpRedeemScript);
+        Script fedInputScript = flyoverP2sh.createEmptyInputScript(redeemData.keys.get(0), redeemData.redeemScript);
+        btcTx.getInput(FIRST_INPUT_INDEX).setScriptSig(fedInputScript);
+
+        Sha256Hash expectedSigHash = btcTx.hashForSignature(
+            FIRST_INPUT_INDEX,
+            flyoverP2shErpRedeemScript,
+            BtcTransaction.SigHash.ALL,
+            false
+        );
+
+        // Act
+        Optional<Sha256Hash> sigHash = BitcoinUtils.getFirstInputSigHash(btcTx);
+
+        // Assert
+        Assertions.assertTrue(sigHash.isPresent());
+        Assertions.assertEquals(expectedSigHash, sigHash.get());
+    }
+
+    @Test
+    void test_getFirstInputSigHash_no_input() {
+        // Arrange
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+        btcTx.addOutput(Coin.COIN, destinationAddress);
+
+        // Act
+        Optional<Sha256Hash> sigHash = BitcoinUtils.getFirstInputSigHash(btcTx);
+
+        // Assert
+        Assertions.assertFalse(sigHash.isPresent());
+    }
+
+    @Test
+    void test_getFirstInputSigHash_invalid_input_no_redeem_script() {
+        // Arrange
+        Address destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+        btcTx.addInput(BitcoinTestUtils.createHash(1), 0, new Script(new byte[]{}));
+        btcTx.addOutput(Coin.COIN, destinationAddress);
+
+        // Act
+        Optional<Sha256Hash> sigHash = BitcoinUtils.getFirstInputSigHash(btcTx);
+
+        // Assert
+        Assertions.assertFalse(sigHash.isPresent());
+    }
+
+    @Test
+    void test_getFirstInputSigHash_many_inputs() {
+        // Arrange
+        int totalSigners = signers.size();
+        int requiredSignatures = totalSigners / 2 + 1;
+
+        Script redeemScript = ScriptBuilder.createRedeemScript(requiredSignatures, signers);
+        Script p2SHScript = ScriptBuilder.createP2SHOutputScript(requiredSignatures, pubKeys);
+
+        Address destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+
+        for (int i = 1; i <= 50; i++) {
+            btcTx.addInput(BitcoinTestUtils.createHash(1), 0, new Script(new byte[]{}));
+            RedeemData redeemData = RedeemData.of(pubKeys, redeemScript);
+
+            Script inputScript = p2SHScript.createEmptyInputScript(redeemData.keys.get(0), redeemData.redeemScript);
+            btcTx.getInput(i-1).setScriptSig(inputScript);
+        }
+        btcTx.addOutput(Coin.COIN, destinationAddress);
+
+        Sha256Hash expectedSigHash = btcTx.hashForSignature(
```

Comment:
We could also assert that the sighash for different inputs is different

Reply:
Done